File: code/ddouble/ocaml/powbench.ml
(* x^100 computed two ways, on the machine Rocq runs on.                    *)
(*                                                                          *)
(* "2 floats" is a double word: the definitions are the ones of dwarith.v,  *)
(* copied over.  In particular the product of two floats is split by        *)
(* Dekker's method, because a primitive float has no fused multiply add.    *)
(*                                                                          *)
(* "3 ints" is a mantissa of 120 bits held in two integers of 60 bits, and  *)
(* an exponent.  Sixty rather than the sixty three an OCaml integer holds:  *)
(* that is a choice, it leaves room to add a few of them without watching   *)
(* the sign bit, and it costs no multiplication.  The product is taken in   *)
(* both shapes, one that throws away the half that does not fit and one     *)
(* that rounds to nearest.                                                  *)

let pow = 100                       (* the power, so 99 multiplications     *)
let rep = 200_000                   (* how many powers a run computes       *)

(* ---------------------------------------------------------------------- *)
(* 2 floats                                                               *)
(* ---------------------------------------------------------------------- *)

type dw = { h : float; l : float }

let c_const = 134217729.            (* 2^27 + 1                             *)

(* splitC, dekker and fastTwoSum written out inside timesDwDw rather than   *)
(* called: a float held in a local name stays in a register, one returned   *)
(* from a function is put in a box.  Both sides then build exactly one      *)
(* record per multiplication, which is what makes the two times comparable. *)
let times_dw_dw x y =
  let xh = x.h and xl = x.l and yh = y.h and yl = y.l in
  (* splitC xh *)
  let gx = c_const *. xh in
  let dx = xh -. gx in
  let xhh = gx +. dx in
  let xhl = xh -. xhh in
  (* splitC yh *)
  let gy = c_const *. yh in
  let dy = yh -. gy in
  let yhh = gy +. dy in
  let yhl = yh -. yhh in
  (* dekker xh yh *)
  let ch = xh *. yh in
  let t1 = -. ch +. xhh *. yhh in
  let t2 = t1 +. xhh *. yhl in
  let t3 = t2 +. xhl *. yhh in
  let cl1 = t3 +. xhl *. yhl in
  (* the three remaining products, then fastTwoSum *)
  let cl2 = xh *. yl +. xl *. yh in
  let cl3 = cl1 +. cl2 in
  let s = ch +. cl3 in
  let z = s -. ch in
  { h = s; l = cl3 -. z }

(* ---------------------------------------------------------------------- *)
(* 3 ints: the value is (hi * 2^60 + lo) * 2^(e - 120), and the mantissa   *)
(* is kept between a half and one, so bit 59 of hi is set                  *)
(* ---------------------------------------------------------------------- *)

let mask30 = (1 lsl 30) - 1
let mask60 = (1 lsl 60) - 1

type ti = { hi : int; lo : int; e : int }

(* The low 60 bits of a product are just a * b: OCaml keeps the low bits    *)
(* exactly.  The high 60 need the schoolbook split, which is where the      *)
(* work is, since no integer here can hold a product of two whole words.    *)
let mulh a b =
  let al = a land mask30 and ah = a lsr 30 in
  let bl = b land mask30 and bh = b lsr 30 in
  let t = al * bl and m = ah * bl and n = al * bh in
  let acc = (t lsr 30) + (m land mask30) + (n land mask30) in
  ah * bh + (m lsr 30) + (n lsr 30) + (acc lsr 30)

(* the three digits of the product that reach the answer, the fourth being  *)
(* too small to carry anything                                              *)
let ti_mul a b =
  let p11h = mulh a.hi b.hi and p11l = a.hi * b.hi land mask60 in
  let p10h = mulh a.hi b.lo and p10l = a.hi * b.lo land mask60 in
  let p01h = mulh a.lo b.hi and p01l = a.lo * b.hi land mask60 in
  let p00h = mulh a.lo b.lo in
  let d1 = p10l + p01l + p00h in
  let d2 = p11l + p10h + p01h + (d1 lsr 60) in
  let d3 = p11h + (d2 lsr 60) in
  let r0 = d2 land mask60 and g = d1 land mask60 in
  if d3 lsr 59 = 1 then { hi = d3; lo = r0; e = a.e + b.e }
  else { hi = (d3 lsl 1) lor (r0 lsr 59);
         lo = ((r0 lsl 1) lor (g lsr 59)) land mask60;
         e = a.e + b.e - 1 }

let ti_mulr a b =
  let p11h = mulh a.hi b.hi and p11l = a.hi * b.hi land mask60 in
  let p10h = mulh a.hi b.lo and p10l = a.hi * b.lo land mask60 in
  let p01h = mulh a.lo b.hi and p01l = a.lo * b.hi land mask60 in
  let p00h = mulh a.lo b.lo in
  let d1 = p10l + p01l + p00h in
  let d2 = p11l + p10h + p01h + (d1 lsr 60) in
  let d3 = p11h + (d2 lsr 60) in
  let r0 = d2 land mask60 and g = d1 land mask60 in
  let sh = 1 - (d3 lsr 59) in       (* one shift, or none                   *)
  let n1 = if sh = 1 then (d3 lsl 1) lor (r0 lsr 59) else d3 in
  let n0 =
    if sh = 1 then ((r0 lsl 1) lor (g lsr 59)) land mask60 else r0 in
  let rb =                          (* the first bit that does not fit      *)
    if sh = 1 then (g lsr 58) land 1 else g lsr 59 in
  let n0 = n0 + rb in
  let n1 = n1 + (n0 lsr 60) in
  let n0 = n0 land mask60 in
  if n1 lsr 60 = 1 then { hi = 1 lsl 59; lo = 0; e = a.e + b.e - sh + 1 }
  else { hi = n1; lo = n0; e = a.e + b.e - sh }

let ti_of_float x =
  let f, e = Float.frexp x in
  { hi = int_of_float (Float.ldexp f 53) lsl 7; lo = 0; e }

(* ---------------------------------------------------------------------- *)

let base k = 3.14159265358979311600 +. 1e-12 *. float_of_int k

(* one power function each, rather than one taking the multiplication as an *)
(* argument, which would put a call through a pointer on every step         *)
let rec ti_pow r x n = if n = 0 then r else ti_pow (ti_mul r x) x (n - 1)
let rec tr_pow r x n = if n = 0 then r else tr_pow (ti_mulr r x) x (n - 1)
let rec dw_pow r x n = if n = 0 then r else dw_pow (times_dw_dw r x) x (n - 1)
let rec fl_pow r x n = if n = 0 then r else fl_pow (r *. x) x (n - 1)

let run_ti () =
  let t = Sys.time () in
  let rec go k acc last =
    if k = rep then (Sys.time () -. t, acc, last)
    else
      let x = ti_of_float (base k) in
      let r = ti_pow x x (pow - 1) in
      go (k + 1) (acc + r.hi) r
  in
  go 0 0 { hi = 0; lo = 0; e = 0 }

let run_tr () =
  let t = Sys.time () in
  let rec go k acc last =
    if k = rep then (Sys.time () -. t, acc, last)
    else
      let x = ti_of_float (base k) in
      let r = tr_pow x x (pow - 1) in
      go (k + 1) (acc + r.hi) r
  in
  go 0 0 { hi = 0; lo = 0; e = 0 }

let run_dw () =
  let t = Sys.time () in
  let rec go k acc last =
    if k = rep then (Sys.time () -. t, acc, last)
    else
      let x = { h = base k; l = 0. } in
      let r = dw_pow x x (pow - 1) in
      go (k + 1) (acc +. r.h) r
  in
  go 0 0. { h = 0.; l = 0. }

let run_fl () =
  let t = Sys.time () in
  let rec go k acc last =
    if k = rep then (Sys.time () -. t, acc, last)
    else
      let x = base k in
      let r = fl_pow x x (pow - 1) in
      go (k + 1) (acc +. r) r
  in
  go 0 0. 0.

let () =
  let w0 = Gc.minor_words () in
  let tt, at, rt = run_ti () in
  let w1 = Gc.minor_words () in
  let tr, ar, rr = run_tr () in
  let w2 = Gc.minor_words () in
  let td, ad, rd = run_dw () in
  let w3 = Gc.minor_words () in
  let tf, af, rf = run_fl () in
  let n = float_of_int rep *. float_of_int (pow - 1) in
  Printf.printf "pi^%d, %d times, %d multiplications each\n\n" pow rep (pow - 1);
  Printf.printf "  3 ints    %6.2f ns/mul   the low half thrown away\n"
    (1e9 *. tt /. n);
  Printf.printf "  3 ints    %6.2f ns/mul   rounded to nearest\n"
    (1e9 *. tr /. n);
  Printf.printf "  2 floats  %6.2f ns/mul\n" (1e9 *. td /. n);
  Printf.printf "  1 float   %6.2f ns/mul\n" (1e9 *. tf /. n);
  Printf.printf "\n  3 ints / 2 floats   %.2f thrown away   %.2f rounded\n"
    (tt /. td) (tr /. td);
  Printf.printf "\nthe last power, mantissa of 120 bits and exponent\n";
  Printf.printf "  3 ints    %015x%015x  2^%d  thrown away\n" rt.hi rt.lo rt.e;
  Printf.printf "  3 ints    %015x%015x  2^%d  rounded\n" rr.hi rr.lo rr.e;
  Printf.printf "  2 floats  %h %h\n" rd.h rd.l;
  Printf.printf "  1 float   %.17e\n" rf;
  Printf.printf "\nwords put in the heap per multiplication  %.1f %.1f %.1f\n"
    ((w1 -. w0) /. n) ((w2 -. w1) /. n) ((w3 -. w2) /. n);
  Printf.printf "\nchecksums %d %d %.17e %.17e\n" at ar ad af

File: code/ddouble/c/powbench.c
/* x^100 computed two ways, timed and compared bit by bit.                  */
/*                                                                          */
/* "3 ints": a 128-bit mantissa held in two 64-bit limbs plus a separate    */
/* exponent, the way a bignum library simulates a floating point number.    */
/* "2 floats": a double word, a pair of binary64 numbers whose unevaluated  */
/* sum is the value, giving 106 significant bits.                           */
/*                                                                          */
/* The 3-int multiplication truncates instead of rounding to nearest: that  */
/* is a choice, it makes the integer side as fast as it can be, and it      */
/* costs it at most one bit of the 128 it carries.                          */

#include <stdio.h>
#include <stdint.h>
#include <math.h>
#include <time.h>

#define POW 100         /* the exponent, so 99 multiplications per power    */
#define REP 1000000     /* how many powers a run computes                   */

/* ---------------------------------------------------------------------- */
/* 3 ints: the value is (hi:lo) * 2^(e-128), with 1/2 <= (hi:lo)/2^128 < 1 */
/* ---------------------------------------------------------------------- */

typedef struct { uint64_t hi, lo; int64_t e; } ti;

static inline ti ti_mul(ti a, ti b)
{
  __uint128_t hh = (__uint128_t) a.hi * b.hi;
  __uint128_t hl = (__uint128_t) a.hi * b.lo;
  __uint128_t lh = (__uint128_t) a.lo * b.hi;
  __uint128_t ll = (__uint128_t) a.lo * b.lo;
  /* the bits of weight below 2^-128 only matter through their carry       */
  __uint128_t mid = (hl & UINT64_MAX) + (lh & UINT64_MAX) + (ll >> 64);
  __uint128_t r   = hh + (hl >> 64) + (lh >> 64) + (mid >> 64);
  uint64_t rhi = (uint64_t) (r >> 64), rlo = (uint64_t) r;
  int64_t e = a.e + b.e;
  /* both factors are below 1, so the product needs at most one shift      */
  if (!(rhi >> 63)) { rhi = (rhi << 1) | (rlo >> 63); rlo <<= 1; e--; }
  ti c = { rhi, rlo, e };
  return c;
}

static ti ti_of_double(double x)
{
  int e; double f = frexp(x, &e);
  uint64_t m = (uint64_t) ldexp(f, 53);
  ti c = { m << 11, 0, e };
  return c;
}

/* ---------------------------------------------------------------------- */
/* 2 floats: the value is hi + lo, the sum left unevaluated               */
/* ---------------------------------------------------------------------- */

typedef struct { double hi, lo; } dw;

static inline dw dw_mul(dw a, dw b)
{
  double p = a.hi * b.hi;
  double e = fma(a.hi, b.hi, -p);        /* the part p dropped, exactly    */
  e += a.hi * b.lo + a.lo * b.hi;
  double s = p + e;
  dw c = { s, (p - s) + e };
  return c;
}

/* the double word read back as a 128-bit mantissa, so the two answers     */
/* can be lined up digit against digit                                     */
static ti ti_of_dw(dw a)
{
  int e1; double f1 = frexp(a.hi, &e1);
  uint64_t m1 = (uint64_t) ldexp(f1, 53);
  __uint128_t s = (__uint128_t) m1 << 75;
  if (a.lo != 0) {
    int e2; double f2 = frexp(fabs(a.lo), &e2);
    uint64_t m2 = (uint64_t) ldexp(f2, 53);
    int sh = 75 - (e1 - e2);
    __uint128_t t = sh >= 0 ? (__uint128_t) m2 << sh : (__uint128_t) m2 >> -sh;
    if (a.lo > 0) s += t; else s -= t;
  }
  int64_t e = e1;
  while (!(s >> 127)) { s <<= 1; e--; }
  ti c = { (uint64_t) (s >> 64), (uint64_t) s, e };
  return c;
}

/* ---------------------------------------------------------------------- */
/* Each run computes LANES powers side by side.  With one lane the timing  */
/* is the latency of a multiplication, every one waiting for the previous  */
/* one; with four lanes the chains are independent and the timing is the   */
/* throughput, what the processor sustains when it has work to overlap.    */
/* ---------------------------------------------------------------------- */

static double check = 0;

static double now(void)
{
  struct timespec t; clock_gettime(CLOCK_MONOTONIC, &t);
  return t.tv_sec + 1e-9 * t.tv_nsec;
}

#define RUN(NAME, LANES, TYPE, OF, MUL, PART)                            \
static double NAME(TYPE *last)                                           \
{                                                                        \
  double t = now(), acc = 0;                                             \
  TYPE r[LANES], x[LANES];                                               \
  for (int k = 0; k < REP; k += LANES) {                                 \
    for (int j = 0; j < LANES; j++) {                                    \
      x[j] = OF(M_PI + 1e-12 * (k + j)); r[j] = x[j];                    \
    }                                                                    \
    for (int i = 1; i < POW; i++)                                        \
      for (int j = 0; j < LANES; j++) r[j] = MUL(r[j], x[j]);            \
    /* an empty asm that claims to read and write every power: it keeps   \
       the chains alive and stops the compiler running several rounds of  \
       the outer loop side by side, which would hide the latency           */ \
    for (int j = 0; j < LANES; j++) __asm__ volatile("" : "+m"(r[j]));    \
    acc += (double) (r[0] PART);                                          \
  }                                                                      \
  t = now() - t;                                                         \
  *last = r[LANES - 1]; check += acc;                                    \
  return t;                                                              \
}

static inline double d_of(double x) { return x; }
static inline double d_mul(double a, double b) { return a * b; }
static inline dw dw_of(double x) { dw c = { x, 0 }; return c; }

RUN(ti_lat, 1, ti, ti_of_double, ti_mul, .hi)
RUN(ti_th2, 2, ti, ti_of_double, ti_mul, .hi)
RUN(ti_thr, 4, ti, ti_of_double, ti_mul, .hi)
RUN(dw_lat, 1, dw, dw_of, dw_mul, .hi)
RUN(dw_th2, 2, dw, dw_of, dw_mul, .hi)
RUN(dw_thr, 4, dw, dw_of, dw_mul, .hi)
RUN(d_lat,  1, double, d_of, d_mul, )
RUN(d_th2,  2, double, d_of, d_mul, )
RUN(d_thr,  4, double, d_of, d_mul, )

int main(void)
{
  ti ri; dw rd; double rp;
  double til = ti_lat(&ri), ti2 = ti_th2(&ri), tit = ti_thr(&ri);
  double dwl = dw_lat(&rd), dw2 = dw_th2(&rd), dwt = dw_thr(&rd);
  double dl  = d_lat(&rp),  d2  = d_th2(&rp),  dt  = d_thr(&rp);
  double n = (double) REP * (POW - 1);

  printf("pi^%d, %d times, %d multiplications each\n\n", POW, REP, POW - 1);
  printf("  ns per multiplication, with 1, 2 and 4 chains side by side\n\n");
  printf("  3 ints     %6.2f  %6.2f  %6.2f\n", 1e9*til/n, 1e9*ti2/n, 1e9*tit/n);
  printf("  2 floats   %6.2f  %6.2f  %6.2f\n", 1e9*dwl/n, 1e9*dw2/n, 1e9*dwt/n);
  printf("  1 float    %6.2f  %6.2f  %6.2f\n", 1e9*dl/n,  1e9*d2/n,  1e9*dt/n);
  printf("\n  3 ints / 2 floats   %.2f    %.2f    %.2f\n",
         til/dwl, ti2/dw2, tit/dwt);

  ti rdi = ti_of_dw(rd);
  printf("\nthe last power, as a 128-bit mantissa and an exponent\n");
  printf("  exact     8dfdc6c2c7764947f43e267c0fffe51e  2^166\n");
  printf("  3 ints    %016lx%016lx  2^%ld\n", ri.hi, ri.lo, (long) ri.e);
  printf("  2 floats  %016lx%016lx  2^%ld\n", rdi.hi, rdi.lo, (long) rdi.e);
  printf("  1 float   %.17e\n", rp);
  printf("\nchecksum %.17e\n", check);
  return 0;
}
